File: src/main/java/jisop/command_line/parse/ArgumentParameter.java
package jisop.command_line.parse;

import jisop.infrastructure.Strings;

import java.util.Arrays;

/**
 *
 * @author mathieu
 */
public class ArgumentParameter {
    public ArgumentParameter(String prototype, String[] names, String delimiter, Integer ordinal) {
        this.prototype = prototype;
        aliases = names;
        this.delimiter = delimiter;
        this.ordinal = ordinal;
    }
    public final String prototype;
    public final Integer ordinal;

    public static ArgumentParameter parse(String value) {
        ArgumentParameter ordinalParameter= OrdinalParameter.tryParse(value);
        if (ordinalParameter!=null)
            return ordinalParameter;
        ArgumentParameter optionParameter= OptionParameter.tryParse(value);
        if (optionParameter!=null)
            return optionParameter;
        ArgumentParameter visualStudioParameter= VisualStudioParameter.tryParse(value);
        if (visualStudioParameter!=null)
            return visualStudioParameter;
        throw new RuntimeException("failed to parse "+value);
    }
    public final String[] aliases;
    public final String delimiter;

    public String help() {
        return "--" + Strings.join(", or ", aliases)
                + (Strings.isNullOrEmpty(delimiter)
                ? ""
                : " " + delimiter);
    }

    @Override
    public String toString() {
        return prototype;
    }

    public boolean hasAlias(String value) {
        return Strings.containsStringIgnoreCase(aliases, value);
    }

    public boolean accept(int index, String val) {
        if (ordinal ==null) {
            return hasAlias(val);
        } else {
            return ordinal.equals(index) && hasAlias(val);
        }
    }
    public boolean accept(String val) {
        return hasAlias(val);
    }

    public String longAlias() {
        return Arrays.asList(aliases)
                .stream()
                .max( (a,b) -> Integer.compare(a.length(),b.length()))
                .orElse(null);
    }
}

File: src/main/java/isop/command_line/parse/ArgumentParameter.kt
package isop.command_line.parse

import isop.infrastructure.Strings

import java.util.Arrays

/**
 *
 * @author mathieu
 */
class ArgumentParameter(val prototype: String, val aliases: Array<String>, val delimiter: String?, val ordinal: Int?) {

    fun help(): String {
        return ("--" + Strings.join(", or ", aliases)
                + if (Strings.isNullOrEmpty(delimiter))
            ""
        else
            " $delimiter")
    }

    override fun toString(): String {
        return prototype
    }

    fun hasAlias(value: String): Boolean {
        return Strings.containsStringIgnoreCase(aliases, value)
    }

    fun accept(index: Int, `val`: String): Boolean {
        return if (ordinal == null) {
            hasAlias(`val`)
        } else {
            ordinal == index && hasAlias(`val`)
        }
    }

    fun accept(`val`: String): Boolean {
        return hasAlias(`val`)
    }

    fun longAlias(): String {
        return Arrays.asList(*aliases)
                .stream()
                .max { a, b -> Integer.compare(a.length, b.length) }
                .orElse(null)
    }

    companion object {

        fun parse(value: String): ArgumentParameter {
            val ordinalParameter = OrdinalParameter.tryParse(value)
            if (ordinalParameter != null)
                return ordinalParameter
            val optionParameter = OptionParameter.tryParse(value)
            if (optionParameter != null)
                return optionParameter
            val visualStudioParameter = VisualStudioParameter.tryParse(value)
            if (visualStudioParameter != null)
                return visualStudioParameter
            throw RuntimeException("failed to parse $value")
        }
    }
}
